File: viewer-web/core/propagate.js
// Padded grid and band-limited angular-spectrum propagator, ported from src/grid.cpp and
// src/propagate.cpp with the same equations:
//   transfer function  H = exp[i 2 pi (z / lambda) sqrt(1 - (lambda fx)^2 - (lambda fy)^2)]
//                      (Goodman 2017 eq. 3-78), evanescent bins (arg <= 0) set to zero (eq. 3-72);
//   band limit         f_lim = 1 / (lambda sqrt((2 z / S)^2 + 1)), rect window
//                      (Matsushima & Shimobaba 2009 eqs. 13-14, S = padded extent);
//   padding            next 7-smooth size >= 2 N (next_smooth).
// Fields are { re, im } Float64Arrays of an n x n grid, [i, j] with i along x.
(function () {
  // Module factory. As a classic script it also registers its own source so that the page can
  // build a Web Worker from it (a worker script cannot be loaded from a file:// URL).
  function factory(root) {
  const fftmod = (typeof module !== 'undefined' && module.exports) ? require('./fft.js') : root.DOE_CORE;
  const { fft2, ifft2 } = fftmod;

  function nextPow2(n) { let m = 1; while (m < n) m *= 2; return m; }
  function nextSmooth(n) {
    if (n <= 1) return 1;
    const limit = 2 * n;
    let smooth = [1];
    for (const f of [2, 3, 5, 7]) {
      const grown = [];
      for (const s of smooth) for (let v = s; v <= limit; v *= f) grown.push(v);
      smooth = grown;
    }
    let best = limit;
    for (const s of smooth) if (s >= n && s < best) best = s;
    return best;
  }

  class Grid {
    constructor(n, pitch, wavelength) { this.n = n; this.pitch = pitch; this.wavelength = wavelength; }
    static padded(active, pitch, wavelength, pad = 2.0) { return new Grid(nextSmooth(Math.ceil(pad * active)), pitch, wavelength); }
    // Picture-clear rule (Grid::padded_for in the C++): S >= D + d tan(theta_max), so light wrapping around the
    // periodic window misses the picture. `round` is nextSmooth (the C++ sizes) or nextPow2 (the browser's fast FFT).
    static paddedFor(active, pitch, wavelength, distance, padMin = 1.0, round = nextSmooth) {
      const theta = Math.asin(Math.min(1, wavelength / (2 * pitch)));
      const required = (active * pitch + distance * Math.tan(theta)) / pitch;
      return new Grid(round(Math.max(1, Math.ceil(padMin * active), Math.ceil(required - 1e-9))), pitch, wavelength);
    }
    k() { return 2 * Math.PI / this.wavelength; }
    extent() { return this.n * this.pitch; }
    x(i) { return (i - this.n / 2) * this.pitch; }
    // numpy.fft.fftfreq order: non-negative bins first, then the negative ones
    f(i) { const half = Math.floor((this.n + 1) / 2); const bin = i < half ? i : i - this.n; return bin / (this.n * this.pitch); }
    maxDiffractionAngle() { return Math.asin(Math.min(1, this.wavelength / (2 * this.pitch))); }
  }

  function transferFunction(grid, distance, bandLimit) {
    const n = grid.n, lam = grid.wavelength, N = n * n;
    const re = new Float64Array(N), im = new Float64Array(N);
    let fLimit = Infinity;
    if (bandLimit && distance !== 0) fLimit = 1 / (lam * Math.sqrt((2 * Math.abs(distance) / grid.extent()) ** 2 + 1));
    let evanescentCount = 0, pass = 0;
    const fs = new Float64Array(n);
    for (let i = 0; i < n; ++i) fs[i] = grid.f(i);
    for (let i = 0; i < n; ++i) {
      const fx = fs[i];
      for (let j = 0; j < n; ++j) {
        const fy = fs[j], k = i * n + j;
        const arg = 1 - (lam * fx) * (lam * fx) - (lam * fy) * (lam * fy);
        if (arg <= 0) { ++evanescentCount; continue; }
        if (Math.abs(fx) > fLimit || Math.abs(fy) > fLimit) continue;
        const ph = 2 * Math.PI / lam * Math.sqrt(arg) * distance;
        re[k] = Math.cos(ph); im[k] = Math.sin(ph); ++pass;
      }
    }
    return { re, im, fLimit, evanescentCount, evanescentFraction: evanescentCount / N, passbandFraction: pass / N };
  }

  function mulInPlace(re, im, hre, him, conj) {
    const s = conj ? -1 : 1;
    for (let k = 0; k < re.length; ++k) {
      const a = re[k], b = im[k], c = hre[k], d = s * him[k];
      re[k] = a * c - b * d; im[k] = a * d + b * c;
    }
  }

  class AngularSpectrum {
    constructor(grid, distance, bandLimit = true) {
      this.grid = grid; this.distance = distance; this.bandLimit = bandLimit;
      const tf = transferFunction(grid, distance, bandLimit);
      this.H = tf; this.evanescentFraction = tf.evanescentFraction; this.passbandFraction = tf.passbandFraction;
    }
    forward(u) { const re = u.re.slice(), im = u.im.slice(), n = this.grid.n; fft2(re, im, n); mulInPlace(re, im, this.H.re, this.H.im, false); ifft2(re, im, n); return { re, im }; }
    adjoint(v) { const re = v.re.slice(), im = v.im.slice(), n = this.grid.n; fft2(re, im, n); mulInPlace(re, im, this.H.re, this.H.im, true); ifft2(re, im, n); return { re, im }; }
    // Planes at the given distances, one forward FFT shared; onPlane(p, {re, im}) receives a
    // buffer that is reused between planes.
    sweep(u, distances, onPlane) {
      const n = this.grid.n, Ure = u.re.slice(), Uim = u.im.slice();
      fft2(Ure, Uim, n);
      const wre = new Float64Array(Ure.length), wim = new Float64Array(Uim.length);
      for (let p = 0; p < distances.length; ++p) {
        const tf = transferFunction(this.grid, distances[p], this.bandLimit);
        wre.set(Ure); wim.set(Uim);
        mulInPlace(wre, wim, tf.re, tf.im, false);
        ifft2(wre, wim, n);
        onPlane(p, { re: wre, im: wim });
      }
    }
  }

  function samplingReport(grid, distance, aperture) {
    const lam = grid.wavelength, theta = grid.maxDiffractionAngle();
    const r = {};
    r.maxAngleRad = theta;
    r.spotSizeM = lam * distance / Math.max(aperture, 1e-12);
    r.spotSizePx = r.spotSizeM / grid.pitch;
    r.signalSpreadM = 2 * distance * Math.tan(theta);
    r.windowExtentM = grid.extent();
    r.fresnelNumber = aperture * aperture / (lam * Math.max(distance, 1e-12));
    r.wrapMissesPicture = aperture + 0.5 * r.signalSpreadM <= r.windowExtentM * (1 + 1e-12);   // D + d tan(theta) <= S: wrapped light misses the picture
    r.spotResolved = r.spotSizePx <= 1;
    r.warnings = [];
    if (!r.wrapMissesPicture) r.warnings.push('light wrapping around the periodic window reaches the picture: enlarge the window, reduce the distance or the pitch');
    if (!r.spotResolved) r.warnings.push('diffraction-limited spot is larger than one target pixel: finer detail is unreachable');
    return r;
  }

  const api = { nextSmooth, nextPow2, Grid, transferFunction, AngularSpectrum, samplingReport };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  else { root.DOE_CORE = root.DOE_CORE || {}; Object.assign(root.DOE_CORE, api); }
  }
  const root = typeof globalThis !== 'undefined' ? globalThis : this;
  factory(root);
  if (!(typeof module !== 'undefined' && module.exports)) (root.DOE_MODULE_SOURCES = root.DOE_MODULE_SOURCES || []).push(factory.toString());
})();

File: src/propagate.cpp
#include "doe/propagate.hpp"

#include "parallel.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <numbers>

namespace doe {

KzMap kz_map(const Grid& grid) {
    const std::size_t n = static_cast<std::size_t>(grid.n);
    const double lam = grid.wavelength;
    KzMap m{Array2<double>(n, n, 0.0), Array2<unsigned char>(n, n, 0)};
    par::for_each(n, [&](std::size_t i) {
        const double fx = grid.f(i);
        for (std::size_t j = 0; j < n; ++j) {
            const double fy = grid.f(j);
            // Goodman eq. (3-78): exponent 2 pi (z / lambda) sqrt(1 - (lambda fx)^2 - (lambda fy)^2)
            const double arg = 1.0 - (lam * fx) * (lam * fx) - (lam * fy) * (lam * fy);
            if (arg <= 0.0) {  // evanescent (eq. 3-72): carries no energy away, dropped
                m.evanescent(i, j) = 1;
                continue;
            }
            m.kz(i, j) = 2.0 * std::numbers::pi / lam * std::sqrt(arg);
        }
    });
    return m;
}

TransferFunction transfer_function(const KzMap& kz, const Grid& grid, double distance, bool band_limit) {
    const std::size_t n = static_cast<std::size_t>(grid.n);
    const double lam = grid.wavelength;
    TransferFunction tf{Field<double>(n, n), kz.evanescent, std::numeric_limits<double>::infinity()};

    // Matsushima & Shimobaba 2009 eq. (13) with du = 1 / S (S = padded extent):
    // f_lim = 1 / (lambda sqrt((2 z / S)^2 + 1)). Skipped for zero distance
    // (H == 1, nothing to alias).
    if (band_limit && distance != 0.0) {
        const double S = grid.extent();
        tf.f_limit = 1.0 / (lam * std::sqrt(std::pow(2.0 * std::abs(distance) / S, 2) + 1.0));
    }

    par::for_each(n, [&](std::size_t i) {
        const double fx = grid.f(i);
        for (std::size_t j = 0; j < n; ++j) {
            const double fy = grid.f(j);
            if (kz.evanescent(i, j)) {
                tf.H(i, j) = 0.0;
                continue;
            }
            if (std::abs(fx) > tf.f_limit || std::abs(fy) > tf.f_limit) {  // eq. (14) rect window
                tf.H(i, j) = 0.0;
                continue;
            }
            tf.H(i, j) = std::polar(1.0, kz.kz(i, j) * distance);
        }
    });
    return tf;
}

TransferFunction transfer_function(const Grid& grid, double distance, bool band_limit) {
    return transfer_function(kz_map(grid), grid, distance, band_limit);
}

template <class T>
AngularSpectrum<T>::AngularSpectrum(const Grid& grid, double distance, bool band_limit)
    : grid_(grid), distance_(distance), band_limit_(band_limit),
      H_(static_cast<std::size_t>(grid.n), static_cast<std::size_t>(grid.n)),
      Hc_(static_cast<std::size_t>(grid.n), static_cast<std::size_t>(grid.n)),
      fft_(static_cast<std::size_t>(grid.n)) {
    const TransferFunction tf = transfer_function(grid, distance, band_limit);
    std::size_t n_evan = 0, n_pass = 0;
    for (std::size_t k = 0; k < tf.H.data.size(); ++k) {
        H_.data[k] = std::complex<T>(tf.H.data[k]);
        Hc_.data[k] = std::conj(H_.data[k]);
        n_evan += tf.evanescent.data[k];
        n_pass += (std::abs(tf.H.data[k]) > 0.0);
    }
    evanescent_fraction_ = static_cast<double>(n_evan) / static_cast<double>(tf.H.data.size());
    passband_fraction_ = static_cast<double>(n_pass) / static_cast<double>(tf.H.data.size());
}

template <class T>
Field<T> AngularSpectrum<T>::forward(const Field<T>& u) const {
    Field<T> U = fft_.forward(u);
    par::for_each(U.data.size(), [&](std::size_t k) { U.data[k] *= H_.data[k]; });
    return fft_.inverse(U);
}

template <class T>
Field<T> AngularSpectrum<T>::adjoint(const Field<T>& v) const {
    Field<T> V = fft_.forward(v);
    par::for_each(V.data.size(), [&](std::size_t k) { V.data[k] *= Hc_.data[k]; });
    return fft_.inverse(V);
}

template <class T>
std::vector<Field<T>> AngularSpectrum<T>::volume(const Field<T>& u, std::span<const double> distances) const {
    const Field<T> U = fft_.forward(u);
    std::vector<Field<T>> planes;
    planes.reserve(distances.size());
    Field<T> W(U.rows, U.cols);
    const KzMap kz = kz_map(grid_);  // the square roots once, the exponentials per plane
    for (double d : distances) {
        const TransferFunction tf = transfer_function(kz, grid_, d, band_limit_);
        par::for_each(U.data.size(), [&](std::size_t k) { W.data[k] = U.data[k] * std::complex<T>(tf.H.data[k]); });
        planes.push_back(fft_.inverse(W));
    }
    return planes;
}

template <class T>
void AngularSpectrum<T>::sweep(const Field<T>& u, std::span<const double> distances,
                               const std::function<void(std::size_t, const Field<T>&)>& on_plane) const {
    const Field<T> U = fft_.forward(u);
    Field<T> W(U.rows, U.cols), plane(U.rows, U.cols);
    const KzMap kz = kz_map(grid_);
    for (std::size_t p = 0; p < distances.size(); ++p) {
        const TransferFunction tf = transfer_function(kz, grid_, distances[p], band_limit_);
        par::for_each(U.data.size(), [&](std::size_t k) { W.data[k] = U.data[k] * std::complex<T>(tf.H.data[k]); });
        fft_.inverse(W, plane);
        on_plane(p, plane);
    }
}

template class AngularSpectrum<float>;
template class AngularSpectrum<double>;

SamplingReport sampling_report(const Grid& grid, double distance, double aperture) {
    const double lam = grid.wavelength;
    const double theta = grid.max_diffraction_angle();
    SamplingReport r{};
    r.max_angle_rad = theta;
    r.spot_size_m = lam * distance / std::max(aperture, 1e-12);
    r.spot_size_px = r.spot_size_m / grid.pitch;
    r.signal_spread_m = 2.0 * distance * std::tan(theta);
    r.window_extent_m = grid.extent();
    r.fresnel_number = aperture * aperture / (lam * std::max(distance, 1e-12));
    // Light leaving the plate edge reaches D/2 + d tan(theta); the periodic transform wraps
    // it to x - S, which misses the picture |x| <= D/2 while D + d tan(theta) <= S.
    r.wrap_misses_picture = aperture + 0.5 * r.signal_spread_m <= r.window_extent_m * (1 + 1e-12);
    r.spot_resolved = r.spot_size_px <= 1.0;
    return r;
}

}  // namespace doe
